### `raw/`
Links to `/mnt/lab_data2/amtseng/att_priors/data/raw/`
- `ENCODE/`
	- Contains ChIP-seq data fetched from ENCODE
	- `encode_tf_chip_experiments.tsv`
		- A metadata file listing various experiments from ENCODE, filtered for those that are TF-ChIPseq experiments, aligned to hg38, and status released (and not retracted, for example)
		- Downloaded with the following command:
			```
			wget -O encode_tf_chip_experiments.tsv "https://www.encodeproject.org/report.tsv?type=Experiment&status=released&assay_slims=DNA+binding&assay_title=TF+ChIP-seq&assembly=GRCh38"
			```
	- `encode_control_chip_experiments.tsv`
		- A metadata file listing various control experiments from ENCODE (i.e. for a particular cell-line, a ChIP-seq experiment with no immunoprecipitation), filtered for those that are aligned to hg38 and have a status of released
		- Downloaded with the following command:
			```
			wget -O encode_control_chip_experiments.tsv "https://www.encodeproject.org/report.tsv?type=Experiment&status=released&assay_slims=DNA+binding&assay_title=Control+ChIP-seq&assembly=GRCh38"
			```
	- `encode_tf_chip_files.tsv`
		- A metadata file listing various ENCODE files, filtered for those that are aligned to hg38, status released, and of the relevant output types (i.e. unfiltered alignments, called peaks, and optimal IDR-filtered peaks)
		- Downloaded with the following commands:
			```
			FIRST=1; for otype in "unfiltered alignments" "peaks and background as input for IDR" "optimal IDR thresholded peaks"; do wget -O - "https://www.encodeproject.org/report.tsv?type=File&status=released&assembly=GRCh38&output_type=$otype" | awk -v first=$FIRST '(first == 1) || NR > 2' >> encode_tf_chip_files.tsv; FIRST=2; done
			```
		- The API for the download of the experiment and files metadata is described [here](https://app.swaggerhub.com/apis-docs/encodeproject/api/basic_search/)
	- The files for each TF are downloaded by `download_ENCODE_data.py`

### `interim/`
Links to `/mnt/lab_data2/amtseng/att_priors/data/interim/`
- Raw data that has been processed to an intermediate form
- `ENCODE/`
	- `binary/`
		- Processed data needed to train binary models
		- `labels/`
			- Labels generated by the peaks from `raw/`, using `generate_ENCODE_binary_labels.sh`
			- Ambiguous regions are given a value of -1
		- `randomized_labels/`
			- Contains data with the labels randomized; this is based on an existing dataset so the relative quantities of positives and negatives are the same
			- Created using `randomize_binary_dataset.py`
		- `config/`
			- Config files for training specific TF datasets
	- `profile/`
		- Processed data needed to train profile models
		- `labels/`
			- Labels generated by the alignments and peaks from `raw/`, using `generate_ENCODE_profile_labels.sh`
			- In addition to bedTools, this script also requires bedGraphToBigWig from UCSC Tools, as well as the [hg38 chromosome sizes with chrEBV included](https://github.com/ENCODE-DCC/encValData/blob/master/GRCh38/GRCh38_EBV.chrom.sizes)
			- Note that we start with the _unfiltered_ alignments from ENCODE and filter them ourselves
				- The filtering process we do is identical, but we keep duplicate reads, because those can be useful for profile prediction
			- In the peaks BEDs that are created, there are 6 columns
				- The first 3 columns are the location of the peak, where the interval is 1kb centered at the peak summit
				- Columns 4-5 denote the original peak location from the optimal peaks BED
				- Column 6 is the location of the summit itself (i.e. the middle of columns 2-3)
		- `randomized_labels/`
		- `tests/`
			- Some smaller files based on DREAM data that serve as simple tests
		- `config/`
			- Config files for training specific TF datasets

